src/instruction.cpp

```diff
@@ -74,14 +74,15 @@ bool operator==(const instruction& i, instruction_ref ref)
     return std::addressof(i) == std::addressof(*ref);
 }
 
-bool instruction::valid(instruction_ref start, bool check_order) const
+bool instruction::valid(const module& m, bool check_order) const
 {
     return valid() && std::all_of(arguments.begin(), arguments.end(), [&](instruction_ref i) {
                auto self = std::find(i->outputs().begin(), i->outputs().end(), *this);
                bool ret  = self != i->outputs().end();
                if(check_order)
                {
-                   ret = ret and (std::distance(start, i) < std::distance(start, *self));
+                   // check arguments for this instruction before this instruction
+                   ret = ret and (std::distance(m.begin(), i) < std::distance(m.begin(), *self));
```

Comment:
This needs to check if the instruction is in the module here. I think we need to add an overload for `has_instruction` to take a pointer.

Reply:
I did something like `if (check_order and m.has_instruction(i))` in a previous commit. The change I made to `module.cpp:validate()` should make it such that check_order is only true if the correct instruction_ref is in the current module.